### ISSUE TITLE
[Feature] Allow custom publish subdirectory

### ISSUE THREAD

**Issue by esakal** [lerna/lerna#901]
Many libraries in the Angular ecosystem publish to NPM from a subdirectory that was cooked during the build process. The process is usually as followed:
- build the library in a subdirectory (i.e `dist` directory)
- copy package.json, license and readme to the `dist` directory
- do some package.json cleanups in the `dist` directory (delete devDepedencies, move dependencies to peerDependencies, remove scripts...)
- npm publish from the `dist` directory.

I understand that it is not a common technique for node.js package developers but i think it is common for web package developers and my libraries uses the same technique..

While embedding `lerna` into our shared packages monorepo - [kaltura-ng](https://github.com/kaltura/kaltura-ng) I read a lot of issues in lerna and googled about this topic. I read carefully the conversation of issue #91 and even used the same subject with my issue. Unless I missed new issues that address this feature it was marked as 'wontfix' with a recommendation to use the 'package.json:files' array instead.

The reasons for using this approach instead of `package.json:files` array are:
- Many known libraries in the Angular ecosystem does it ([angular/angular](https://github.com/angular/angular), [ReactiveX/rxjs](https://github.com/ReactiveX/rxjs), [cyclejs/cyclejs](https://github.com/cyclejs/cyclejs)) so they must have a reason.
- There is a lot of hoo-ha/complexity with the way node.js resolve modules  for the **web projects** since during the bundling process you **must** refer to the same instance of the library. Unless the bundler (typescript, webpack etc...) provide a hack/workaround/solution to force the library to use its' own node_modules, it will not work. Publishing from sub-directory works just because the `dist/node_modules` not exists.
- During development if the symlink is done against the root, when you import nested class which was not exported in the main index, you will need to refer to the `dist` as part of the path `import { something } from 'my-package/dist/something`. but once you publish from `dist` folder directly, you should somehow fix the pass by removing the `dist` during the tranpiling which is not a valid option.
- The libraries being used as dependencies during development should be assigned as peerDependencies at runtime because you want the application to provide them.

so to recap, we cannot just publish the package with a `dist` folder, we need to publish from `dist` directly

There are some caveats that I could think about with my suggested approach:
1. The dist folder must exists with a package.json inplace before the `lerna bootstrap` process symlink the folders. 
2. The build process should not delete the 'dist' folder, instead it should just clear its content otherwise the symlink of dependent libraries will be broken.

IMO those two caveats are manageable as:
1. we can use a preinstall script to create the folder and a simple package.json file (with at least 'id','name').
2. the build scripts should clear the folder content instead of `rm -rf` the folder itself.
 
## Expected Behavior

when bootstrapping/publishing a package the package.json is being queried for the following config:

```
"config" : {
	"npmDistDirectory" : "dist"
},
```

if this config exists, it will symlink to that folder during bootstrap command and will publish from that folder during  publish command

using the 'config' attribute allow using the same configuration both in `lerna` and in other [node scripts](https://docs.npmjs.com/files/package.json#config).
 
I already modified the 'bootstrap' command in a fork [esakal/lerna](https://github.com/esakal/lerna). I didn't create a PR yet because I'm missing the 'publish' command. I will be happy to continue my work if you are going to consider this feature.

You can see it in action in our repo -  [kaltura-ng](https://github.com/kaltura/kaltura-ng) :
```
$ git clone https://github.com/kaltura/kaltura-ng.git
$ yarn
$ npm run setup
$ cd kaltura-ui/node_modules/@kaltura-ng
$ ll # you should see a symlink to kaltura-common/dist and kaltura-client/dist
```
**NOTE** - your yarn version should be 0.24.6 and above

## Current Behavior

n/a

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 2.0.0-rc.5 |
| `yarn --version` |  v0.24.6 |

| OS | Version |
| --- | --- |
| macOS El Capitan | 10.11.6 |

**Comment by mickaelvalmier** [comment]
Hi,

I have a same problem in a project developed with a lot of Angular 4 libraries.
If dist folder cant' be symlinked directly, the services don't be injected correctly.

I dislike this way proposed by angular (define a package in dist). But if customization of package dir can't be defined, Lerna can't be used with an angular project.

**Comment by captainkovalsky** [comment]
up

**Comment by xealot** [comment]
I recently tried using Lerna for the first time and immediately ran into this issue. Any heavy Typescript development is going to get hamstrung by the omission of this NPM feature.

**Comment by alan-agius4** [comment]
Same issue here I ran into this problem, would love to get this feature, I am even ready to do a PR.

@evocateur any thoughts?

**Comment by alan-agius4** [comment]
Have the same problem here, I’d love to have this feature, and I am willing to do a PR. @evocateur what are your thoughts?

**Comment by Bratn** [comment]
As far as I know this is the only way to accomplish "flat" package structure (without the `/dist` or `/lib` section in the import path) unless skipping `src`/`dist` directories entirely, which isn't practicable if your source requires transpilation.

Let's say I want to import only the `tinyUtil` module from the `utils` namespace inside my `big-obese-package`, what I would like it to look like is:
```javascript
import tinyUtil from 'big-obese-package/utils/tiny-util';
```
If the source is ts/jsx/esnext etc, the recommended way of distributing the package is to transpile it to a `dist` directory ("npmDistDirectory" in @esakal's proposal), include `package.json` and publish. This doesn't seem to be possible with lerna at this moment, which is a pity. 

Instead I will need to transpile it to `./dist`, reference it from the `files` prop in `package.json` and then accept the following import:
```javascript
import tinyUtil from 'big-obese-package/dist/utils/tiny-util';
```
After spending precious time on your API design, naming convention etc, this is a bitter tradeoff.

**Comment by evocateur** [comment]
I don’t understand why this “flat” package structure is better than just another npm package. “big-obese-monolith” packages are an anti-pattern in npm. If you want to expose submodules directly, extract them into a separate package. *Literally* the reason Lerna was created in the first place.

**Comment by Bratn** [comment]
@evocateur: I will not go into argument whether x is better than y - most often I am wrong. But if we ignore my poor choice of lorem-ipsum name (we can call it something else, such as "tiny-lodash"), I can't really see why lerna should hinder the author from using a nested structure within a package and at the same time provide "semantic imports". 

I came here to evaluate lerna as _a tool for managing JavaScript projects with multiple packages_, not as a tool to limit me on how to structure the internals of my packages (whether good or bad).

**Comment by evocateur** [comment]
In my experience, as well as observation of community packages over many years, coupling consumption of a given export to the literal directory structure of a tarball is an extremely hostile anti-pattern. Especially nowadays with ES module exports and whatnot combined with tree-shaking module bundlers, there’s really no fundamental _necessity_ for false-basedir publishing.

Lerna is designed around the way npm works. Packages are published from the same directory as the package.json, and construct their tarball from metadata contained therein. Publishing from a different directory with a modified dependency tree is not idiomatic npm, and lerna will not support it.

**Comment by alan-agius4** [comment]
This will be a major limitation for TypeScript and Angular developers, and in fact some people using lerna with typescript had to do their own publish, or patch the existing implementation.  I myself want to publish from a subdirectory which does contain a `package.json` for several reasons;

1) I don't want to transpile my TS in the `src` folder because I'll end up having a messy file system. `.ts`, `.metadata.json`, `.js`, `.d.ts` all next to each other, and than to `clean` after build will be a total mess.
2) I want to following Google's Angular Package format https://docs.google.com/document/d/1CZC2rcpxffTDfRDs6p1cfbmKNLA6x5O-NtkJglDaBVs/preview
3) Seperation of concerns, why should I have my `dist` files within the same folder of the `src`?

My folder structure
```bash
-- my-lib
---- package.json
---- src
------ index.ts
----dist
------ package.json
------ index.d.ts
------ index.metadata.json
------ esm2015
-------- index.js
------ esm5
-------- index.js
------ bundles
-------- index.js

So in reality i can publish from a subfolder as I do have a `package.json`

**Comment by joewoodhouse** [comment]
Author of several typescript modules here who has hit the exact same issue, all my modules publish from a `dist` folder. Wish I'd known this before starting to use lerna! Would love this feature to be implemented, but will probably have to switch to something else instead now.

**Comment by evocateur** [comment]
I fail to see what typescript has to do with this non-idiomatic subdirectory pattern. You can publish npm modules with transpiled code and typings under `dist` just fine, no mangling of package.json required.

**Comment by alan-agius4** [comment]
That won’t  work if you’d need secondary entry points like
@scope/my-lib/testing

**Comment by austin94** [comment]
For those that are still trying to get "flat-pack" imports working `import Bar from 'foo/Bar'` I was able to solve it by leveraging the `preversion`, `version`, and `postversion` scripts that were added in https://github.com/lerna/lerna/issues/774

For me I got it working by:
- Disabling NPM publishing during the lerna publish command
  - `lerna publish --skip-npm`
- Performing my linting and build during the `preversion` NPM script (or before) which would be built into a `dist` folder
- Copy necessary files into the `dist` folder during the `postversion` NPM script and then call `npm publish dist`. This is handled in a gulp file but I've simplified below.
  - The `package.json` version has been bumped prior to the `postversion` script being called
  - `postversion` is only called if you do not include the `--skip-git` command to lerna publish
  - `"postversion": "cp package.json dist && npm publish dist"`

I still think a config option for `npmDistDirectory` could be beneficial since it would allow a consumer to leverage the `prepublish` lifecycle-hook which seems a little more intuitive.

**Comment by sandangel** [comment]
that's why angular repos not using lerna

**Comment by yordis** [comment]
@austin94 that is painful 💔

**Comment by yordis** [comment]
@evocateur in my case is not because of Angular or because I want to have a fat package but the fact that I can't just export everything throw a single file (pkg.main) because that will work as long as you do not have any naming collision.

I would be great if  in the deployment at least I could specify which folder to actually publish to npm. For the simple reason that you will no need to be adding `/lib` or `/dist` or what have you.

It is not necessary this to be agains anything related to `npm` but the lack of changes on npm itself. 2018 most of the packages uses specific folder for the compiled version of the package, webpack does it, babel community suggest it, parcel does it, rollup community suggest it.

So why could just see the need of this? 

Why we couldn't have specific folder where you will have the final npm package `package.json + code` and we could avoid any weird name on the file path of the package?

**Comment by yordis** [comment]
@evocateur btw, I do not see this related to Typescript only but any package that requires a build process and normally they will put the final package inside specific folder.

**Comment by evocateur** [comment]
That’s literally what the `main` field is for. If you transpile into `dist`, and you have `index.js` as your entry, `”main”: “dist/index.js”` works perfectly fine in your package.json file.

**Comment by evocateur** [comment]
> I can't just export everything [through] a single file (pkg.main) because that will work as long as you do not have any naming collision.

Your packages are too big if that is the case. Embrace the patterns of the npm ecosystem.

**Comment by noherczeg** [comment]
@yordis 

"I would be great if in the deployment at least I could specify which folder to actually publish to npm"

You can tell npm what folders/files to put inside a package via the [files](https://docs.npmjs.com/files/package.json#files) param. I never tried it for sub-directories, but if I got what you are aiming at this could be enough to solve the mentioned problem right?

**Comment by yordis** [comment]
> Your packages are too big if that is the case. Embrace the patterns of the npm ecosystem.

Based on what you can say that?

```js
// factory1.js

export const perform = () => {
  return 'factory1'
}
```

```js
// factory2.js

export const perform = () => {
  return 'factory2'
}
```

```js
// main.js

export * from './factory1'
export * from './factory2'
```

because you can't do the previous example you will end up requiring specific file

```js
// somefile.js

import * as Something from './libname'

console.log(
  Something.perform()
)
```

So, based on what you can assume that you are right saying that the package is too big? 

What will be your argument that I have to pay attention to re-exporting on the main file?

Well, I wouldn't re-export anything if that is your case, the whole point of having files that behave as a module scope in NodeJS.

So tell me how would you resolve this issue please.

**Comment by yordis** [comment]
> Embrace the patterns of the npm ecosystem.

What do you mean by that? Having one function package so you will end up with 100 dependencies when they can live all together in the same package? 

Before Tree-shaking exists, probably makes sense because you didn't want to export the whole thing but now the tooling is advance enough for stopping with such of practices.

**Comment by austin94** [comment]
Hi @yordis :wave: I think there might be a few different conversations going on at once here.

> I would be great if in the deployment at least I could specify which folder to actually publish to npm

Expanding on what @noherczeg mentioned, this is possible with the `main` and/or `files` options within `package.json`

```
// example package.json
...
main: 'dist/index.js',
files: ['dist'],
...
```

Let say your files were all under a `dist` folder. A consumer could retrieve the files in the following way:

```
import * as Everything from 'package';
import * as FactoryOne from 'package/dist/FactoryOne';
import * as FactoryTwo from 'package/dist/FactoryTwo';
etc.
```

There is no _technical_ difference between the code above and 

```
import * as Everything from 'package';
import * as FactoryOne from 'package/FactoryOne';
import * as FactoryTwo from 'package/FactoryTwo';
```

which is what I was looking for. The only reason I was looking for this was to have a "cleaner" import structure.

**Comment by yordis** [comment]
@austin94 right now what I am trying to do is removing `dist` from the path indeed.

**Comment by yordis** [comment]
@austin94 

> Expanding on what @noherczeg mentioned, this is possible with the main and/or files options within package.json

Yes, I am aware of that but that but when you use a folder it will just copy the file inside the folder name so `files: ["lib"]` it will put the files inside `lib`

> There is no technical difference between the code above and

Ones comes from `lib` and the other comes from the root of the project. 

Ideologically, yes you are right, there is not differences.

Technically,

There is a different indeed, in fact, I could have `package/FactoryOne` and `package/lib/FactoryOne` in the same project and have for whatever reason differences between each other.

But I am seeking the same you are seeking, have a clear importing structure.

**Comment by evocateur** [comment]
If anyone wants this, fork lerna and do it yourself. This conversation is over.
